core/src/main/java/jenkins/tasks/SimpleBuildStep.java

```diff
@@ -80,9 +82,38 @@
      * @param listener a place to send output
      * @throws InterruptedException if the step is interrupted
      * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @deprecated Use {@link #perform(Run, FilePath, EnvVars, Launcher, TaskListener) instead.}
      */
-    void perform(@NonNull Run<?,?> run, @NonNull FilePath workspace, @NonNull Launcher launcher,
-                 @NonNull TaskListener listener) throws InterruptedException, IOException;
+    @Deprecated
+    default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull Launcher launcher,
+                         @NonNull TaskListener listener) throws InterruptedException, IOException {
+        // No additional environment available; just use that from the Run.
+        this.perform(run, workspace, run.getEnvironment(listener), launcher, listener);
+    }
+
+    /**
+     * Run this step.
+     * @param run a build this is running as a part of
+     * @param workspace a workspace to use for any file operations
+     * @param env environment variables applicable to this step
+     * @param launcher a way to start processes
+     * @param listener a place to send output
+     * @throws InterruptedException if the step is interrupted
+     * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @since TODO
+     */
+    default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull EnvVars env, @NonNull Launcher launcher,
+                         @NonNull TaskListener listener) throws InterruptedException, IOException {
+        // If this is called, this must be an implementer of the previous API, in which case we call that, discarding
+        // the environment we were given.
+        // But for that to work, that API method must have been implemented.
+        if (Util.isOverridden(SimpleBuildStep.class, this.getClass(),
```

Comment:
@jglick @oleg-nenashev this requires `perform` method not to be final and breaks compatibility with https://github.com/uber/phabricator-jenkins-plugin/pull/344

Reply:
@Zastai certainly not intentional. [JENKINS-62723](https://issues.jenkins-ci.org/browse/JENKINS-62723)

Reply:
I suspect https://github.com/jenkinsci/jenkins/pull/1804 was incorrect.